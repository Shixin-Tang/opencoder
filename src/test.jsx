import React, { useState, useEffect } from "react"
import { render, Box, Text } from "ink2"
import Static from "./components/static.js"
import { render as renderInk } from "../test/testutil"

const Example = () => {
  const [tests, setTests] = useState([])

  useEffect(() => {
    let completedTests = 0
    let timer

    const run = () => {
      // Fake 10 completed tests
      if (completedTests++ < 10) {
        setTests((previousTests) => [
          ...previousTests,
          {
            id: previousTests.length,
            title: `Test #${previousTests.length + 1}`,
          },
        ])

console.log(renderInk(<Text color="green">✔Hello</Text>).lastFrame())

        timer = setTimeout(run, 100)
      }
    }

    run()

    return () => {
      clearTimeout(timer)
    }
  }, [])

  return (
    <Box flexDirection="column" gap={1} position="relative">
      {/* This part will be rendered once to the terminal */}
      <Static items={tests}>
        {(test) => (
          <Box key={test.id}>
            <Text color="green">✔ {test.title}</Text>
          </Box>
        )}
      </Static>

      {/* This part keeps updating as state changes */}
      <Box marginTop={1}>
        <Text dimColor>Completed tests: {tests.length}</Text>
      </Box>
    </Box>
  )
}

render(<Example />)
