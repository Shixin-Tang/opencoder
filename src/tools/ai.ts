import type { ReactNode } from "react"
import { type JSONValue, type Tool, type ToolInvocation } from "ai"

// TODO: annotate message with tool output params, eg file patch, useful for persistent chat messages.

type Streamable = ReactNode | Promise<ReactNode> | JSONValue

export type Renderer<T extends Array<any>> = (...args: T) => AsyncGenerator<Streamable, void, void>

export type CoderTool = Omit<Tool, "execute"> & {
  generate?: Renderer<any>
  renderTitle?: (part: ToolInvocation) => ReactNode
  render?: (part: ToolInvocation) => ReactNode
  execute?: Tool["execute"]
}

export function defineTool(params: CoderTool): CoderTool {
  return params
}
