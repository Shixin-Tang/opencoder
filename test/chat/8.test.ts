import { convertArrayToReadableStream, MockLanguageModelV1 } from "ai/test"

import { buildComponentTree } from "../utils/debugger.js"
import { queryComponentTree } from "../utils/query.js"
import { waitNextRender } from "../utils/render.js"
import { setupTestEnvironment } from "./util.js"
import { delay } from "../utils/delay"

test("user input with special characters is handled correctly", async () => {
  const mockModel = new MockLanguageModelV1({})
  const doStreamMock = vi.fn()
  mockModel.doStream = doStreamMock

  const specialInput = "Line 1\nLine 2\twith tab\nQuotes: \" ' `\nBrackets: [] {}"
  const expectedResponse = [{ type: "text-delta", textDelta: "Received special characters." }]

  doStreamMock.mockImplementationOnce(async (options) => {
    const messages = options.prompt
    const userMessage = Array.isArray(messages)
      ? messages.find((m) => m.role === "user")
      : undefined
    expect(
      userMessage,
      `User message not found in received messages: ${JSON.stringify(messages)}`,
    ).toBeDefined()
    expect(userMessage?.content[0]?.text).toBe(specialInput)

    return {
      stream: convertArrayToReadableStream([
        ...expectedResponse,
        { type: "finish", finishReason: "stop", usage: { promptTokens: 20, completionTokens: 5 } },
      ]),
      rawCall: { rawPrompt: JSON.stringify(messages), rawSettings: {} },
      type: "finish",
      finishReason: "stop",
      usage: { promptTokens: 20, completionTokens: 5 },
    }
  })

  const { instance, stdin, stdout, fiber } = await setupTestEnvironment({ model: mockModel })
  assert(stdin)

  // 2. Simulate user input containing special characters
  stdin.emit("input", specialInput)
  await delay(50)
  await waitNextRender()

  // 3. Verify the input is displayed correctly in the UI *before* sending
  const treeBeforeSend = buildComponentTree(fiber!.current.child)
  const textInputNode = queryComponentTree(treeBeforeSend, "TextInput")
  expect(textInputNode).not.toBeNull()
  expect(textInputNode?.props?.value).toBe(specialInput)
  expect(stdout.get()).toMatchSnapshot("special chars - input displayed")

  // Simulate Send
  stdin.emit("input", "\r")
  await waitNextRender()

  // 5. Wait for model response and check rendering
  await vi.waitFor(
    () => {
      const tree = buildComponentTree(fiber!.current.child)
      if (queryComponentTree(tree, "Spinner")) {
        throw new Error("Spinner still rendering")
      }
      expect(stdout.get()).toContain("Received special characters.")
    },
    { timeout: 1000 },
  )

  expect(doStreamMock).toHaveBeenCalledTimes(1)
  expect(stdout.get()).toMatchSnapshot("special chars - final state")

  instance.unmount()
})
