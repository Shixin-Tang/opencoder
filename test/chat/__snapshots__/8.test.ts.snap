// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user input with special characters is handled correctly > special chars - final state 1`] = `
"> Line 1
Line 2	with tab
Quotes: " ' \`
Brackets: [] {}
▼
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │Received special characters.                                                                     │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·25 tokens"
`;

exports[`user input with special characters is handled correctly > special chars - input displayed 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ > Line 1                                                                                         │
│   Line 2	with tab                                                                                │
│   Quotes: " ' \`                                                                                  │
│   Brackets: [] {}                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;
